### ISSUE TITLE
Improve mass.off.channels.calibrateFromPlan

### ISSUE THREAD

**Issue by joefowler** [usnistgov/mass#183]
**[Original report](https://bitbucket.org/joe_fowler/mass/issues/183) by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

Existing code: [https://bitbucket.org/joe\_fowler/mass/src/3809e9291b81cd640186a389d24aa8d5ef1fde80/src/mass/off/channels.py#lines-729](https://bitbucket.org/joe_fowler/mass/src/3809e9291b81cd640186a389d24aa8d5ef1fde80/src/mass/off/channels.py#lines-729)

In the existing code `ds.calibrationPlan` has specific attribute it goes with, normally `filtValue`. Anytime you define or update `ds.calibrationPlan` it updated the recipe for `energyRough` to be calculated from the chosen attribute. When you call `calibrateFollowingPlan` it fits in `energyRough` space. So it can only be used to accuratly calibrate the chosen attribute, not another attribute like say `filtValueDC`

1. be able to calibrate any `filtValueX` using the same plan
2. to add a two step calibration procedure so the final calibration quality is more independent of the quality of points in `ds.calibrationPlan`

‌

More thoughts:

1. API roughly like

```python
ds.calibrationPlanInit() # note i havent given it a specific filt value
# our corrections are typically small enough that a peak located with filtValue can
# be located with the same value in filtValueDC
ds.calibrationPlanAddPoint("MnKAlpha",state="A", ph=4700)
ds.calibrationPlanAddPoint("TiKAlpha",state="B", ph=3000)
ds.calibrateFollowingPlan("filtValue", calibratedName="energy")
ds.calibrateFollowingPlan("filtValueDC", calibratedName="energyDC")
ds.plotHist("energy", bin_edges)
ds.plotHist("energyDC", bin_edges)
```

2\. Calibration steps 

* make an intermediate calibration: fit lines in energy space using the knowledge in calibration plan to make a calibration \(how exactly this is implemented is up in the air\)
* make a refined calibration: fit lines in energy space using the intermediate calibration

‌

The main questions I have are

1. Should the two step calibration be mandatory or optional?
2. How would one inspect either the intermediate or the final calibration fits?
3. Should the intermediate calibration be added to `ds.recipes`? On the one hand yes, its simpler, on the other hand, it becomes cluttered. Does clutter there matter?
4.  Currently `calibrateFollowingPlan` is implemented with calls to `ds.linefit(“energyRough”,…)`, which is quite convenient and re-uses much code. How can I imlement it to work for any `filtValueX` while still using calls to `ds.linefit`? Or is that not a good goal?

‌

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

I’m thinking it makes sense to let this wait until I need to do a really good calibration with off files, and possibly to combine it with a switch to `MLEModel` instead of mass fitters.

**Comment by joefowler** [comment]
**Original comment by Galen O'Neil (Bitbucket: [oneilg](https://bitbucket.org/oneilg), GitHub: [oneilg](https://github.com/oneilg)).**

----------------------------------------

fixed by PR #119
